File: dongha/week4/376.cpp
#include <iostream>
#include <algorithm>  //min, max함수 쓰기 위함
using namespace std;

// * 전역변수
// dp테이블 (기본적으로 C++ 전역변수는 0으로 초기회)
int dpTable[501][502];

// Java의 nextInt()같은식으로 코드 짜려고 만듦
int getInt() {
    int retValue;
    cin >> retValue;
    return retValue;
}

int main(void) {
    // 삼각형의 크기 n을 입력받음
    const int CONST_N = getInt();
    // 최댓값 구하기용 변수
    int maxValue = 0;
    // DP 진행
    for(int y=1; y<=CONST_N; y++) {
        for(int x=1; x<=y; x++) {
            dpTable[y][x] = max({dpTable[y-1][x], dpTable[y-1][x-1]}) + getInt();
            // 최댓값 갱신
            maxValue = maxValue > dpTable[y][x] ? maxValue : dpTable[y][x];
        }
    }
    cout << maxValue;
    return 0;
}

File: dongha/week4/376.py
def main():
    # dpTable : 가로502 세로501 사이즈 dp 배열 생성 (0번째 행 전체, 그리고 각 행의 좌우 끝은 더미)
    dpTable = [[0]*502 for _ in range(501)]
    # N 입력받음
    CONST_N = int(input())
    # 최댓값 구하기 용 변수
    maxValue = 0 # 값의 범위는 0 이상 9999 이하

    for y in range(1, CONST_N+1):
        intValueIterator = map(int, input().split(' '))
        for x in range(1, y+1):
            # dp 실행
            dpTable[y][x] = max(dpTable[y-1][x], dpTable[y-1][x-1]) + next(intValueIterator)
            # 최댓값 갱신
            maxValue = maxValue if maxValue > dpTable[y][x] else dpTable[y][x]
    print(maxValue)
#####
main()